### ISSUE TITLE
[Swift 5.0] changes the behavior of SourceKit

### ISSUE THREAD

**Issue by norio-nomura** [realm/SwiftLint#2542]
### Describe changes

Since Swift 5.0, SourceKit returns structures inside conditional compilation blocks which never been compiled.
Swift 4.2.1:
```terminal.sh-session
$ TOOLCHAINS=org.swift.42120181030a sourcekitten structure --text '
#if swift(>=6)
let i = 1
#endif
'
{
  "key.diagnostic_stage" : "source.diagnostic.stage.swift.parse",
  "key.length" : 33,
  "key.offset" : 0,
  "key.substructure" : [

  ]
}
```
swift-5.0-DEVELOPMENT-SNAPSHOT-2019-01-04-a:
```terminal.sh-session
$ TOOLCHAINS=org.swift.5020190104a sourcekitten structure --text '
#if swift(>=6)
let i = 1
#endif
'
{
  "key.diagnostic_stage" : "source.diagnostic.stage.swift.parse",
  "key.length" : 33,
  "key.offset" : 0,
  "key.substructure" : [
    {
      "key.accessibility" : "source.lang.swift.accessibility.internal",
      "key.kind" : "source.lang.swift.decl.var.global",
      "key.length" : 9,
      "key.name" : "i",
      "key.namelength" : 1,
      "key.nameoffset" : 20,
      "key.offset" : 16
    }
  ]
}
```

### Impact of changes

- This causes false positive of `legacy_hashing` violation:
```terminal.sh-session
$ echo 'struct MyStruct { var i: Int }
extension MyStruct: Hashable {
#if swift(>=4.2)
    func hash(into hasher: inout Hasher) { hasher.combine(i) }
#else
    var hashValue: Int { return i.hashValue }
#endif
}'|TOOLCHAINS=org.swift.5020190104a swiftlint lint --use-stdin --no-cache
Loading configuration from '.swiftlint.yml'
<nopath>:6:5: warning: Legacy Hashing Violation: Prefer using the `hash(into:)` function instead of overriding `hashValue` (legacy_hashing)
Done linting! Found 1 violation, 0 serious in 1 file.
```

- This affects [how SwiftLint detects the version of SourceKit](https://github.com/realm/SwiftLint/blob/ea171fbbf24c2d099ecc7ebf4c76da8638ef1bae/Source/SwiftLintFramework/Models/SwiftVersion.swift#L38-L92).

**Comment by marcelofabri** [comment]
BTW, I'm less concerned about the `legacy_hashing` false positives than the other item as we could easily deal with that.

Areas where we use `SwiftVersion`:

* Cache
If we can't recognize the version properly, we won't be able to reset the cache between Swift versions

* To check if certain SourceKit features are present in rules
This wouldn't be a problem immediately as all the features we check for would be available on Swift 5, but we need to think how to deal with it in the future.

* Check for language features
I don't think we actually do a good job here, but we probably should. Rules like `legacy_hashing` currently trigger even if the new API is not available. This would be a place where using the compiler version could be useful.

I've created https://bugs.swift.org/browse/SR-9630 to ask for a new request providing the compiler version so we can deal with this in a proper way in the future.

**Comment by jpsim** [comment]
Thanks for filing that @marcelofabri. I agree with the direction.

**Comment by marcelofabri** [comment]
Closing this because we've workaround this for Swift 5.0 and integrated the new SourceKit request for 5.1.
